examples/pennylane_universal_classifier.py

```diff
@@ -0,0 +1,474 @@
+
+.. figure:: ../../examples/figures/universal_dnn.png
+   :scale: 35%
+   :alt: DNN
+
+"Talk is cheap. Show me the code. - Linus Torvalds"
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+"""
+
+import pennylane as qml
+from pennylane import numpy as np
+from pennylane.optimize import AdamOptimizer, GradientDescentOptimizer
+
+import matplotlib.pyplot as plt
+
+
+# Make a dataset of points in and out of a circle
+def circle(samples, center=[0.0, 0.0], radius=0.7):
+    """
+    Generates a dataset of points with 1/0 labels inside a given radius. 
+    
+    Args:
+        samples (int): number of samples to generate.
+        center (tuple): center of the circle
+        radius (float: radius of the circle.
+
+    Returns:
+        Xvals (array of tuples): coordinates of points
+        Xvals (array of ints): classification labels
+    """
+    Xvals, yvals = [], []
+
+    for i in range(samples):
+        x = 2 * (np.random.rand(2)) - 1
+        y = 0
+        if np.linalg.norm(x - center) < radius:
+            y = 1
+        Xvals.append(x)
+        yvals.append(y)
+    return np.array(Xvals), np.array(yvals)
+
+
+def plot_data(x, y, fig=None, ax=None):
+    """
+    Plot data with red/blue values for a binary classification.
+    
+    Args:
+        x (array of tuples): array of data points as tuples
+        y (array of ints): array of data points as tuples
+    """
+    if fig == None:
+        fig, ax = plt.subplots(1, 1, figsize=(5, 5))
+    reds = y == 0
+    blues = y == 1
+    ax.scatter(x[reds, 0], x[reds, 1], c="red", s=20, edgecolor="k")
+    ax.scatter(x[blues, 0], x[blues, 1], c="blue", s=20, edgecolor="k")
+    ax.set_xlabel("$x_1$")
+    ax.set_ylabel("$x_2$")
+
+
+Xdata, ydata = circle(500)
+fig, ax = plt.subplots(1, 1, figsize=(4, 4))
+plot_data(Xdata, ydata, fig=fig, ax=ax)
+plt.show()
+
+
+# Define output labels as quantum state vectors
+def density_matrix(state):
+    """Calculates the density matrix representation of a state.
+
+    Args:
+        state (array[complex]): array representing a quantum state vector
+
+    Returns:
+        dm: (array[complex]): array representing the density matrix.
+    """
+    return state * np.conj(state).T
+
+
+label_0 = [[1.0 + 0.0j], [0.0 + 0.0j]]
+label_1 = [[0.0 + 0.0j], [1.0 + 0.0j]]
+state_labels = [label_0, label_1]
+
+
+# Simple classifier with data reloading and fidelity loss
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+dev = qml.device("default.qubit", wires=1)
+# Use your own pennylane-plugin to run on some particular backend
+
+
+@qml.qnode(dev)
+def qcircuit(var, x=None, y=None):
+    """A variational quantum circuit representing the Universal classifier.
+
+    Args:
+        var (array[float]): array of variables
+        x (array[float]): single input vector
+        y (array[float]): single output state density matrix
+
+    Returns:
+        float: fidelity between output state and input
+    """
+    for v in var:
+        qml.Rot(*x, wires=0)
```

Comment:
In the paper the version for which universality is proven feeds each feature in as w*x+b, or not?

Reply:
Yes. I changed it now but in principle the alternating application of data unitaries and weight unitaries should work. I am also not trying to exactly follow their paper and reproduce their code but more taking the idea/inspiration. For example, they define the cost as sum of fidelities. I have another version where I take those fidelities and define some kind of a cross entropy loss (multiple class). Thanks for the comment though.